### PR TITLE
More Dog Breeds - Now With Puppies Edition

### PR DESCRIPTION
**Pull request by DracoGriffin**
Original: #23813 

- Removed special dog attacks (I couldn't get them to work properly, and don't have the ability to code the attacks)
- Added puppies for each breed.
- Added biosigs for predator spawning.
- Adjusted Bulldog from English Bulldog stats to American Bulldog stats.
- Increased difficulty to reflect strength of canines more accurately (and generally in packs making them more difficult).
- Changed canine sizes as appropriate.
- Added two rare breeds (Rottweiler and Great Pyrenees) that do not spawn except if specified in locations (currently none).
- Added Australian Cattle Dog is it was technically the tenth most common breed in New England, not American Bulldog (was the eleventh) but didn't want to remove it.

Breeding and biosig values could be wildly off, but the intention for breeding is dogs do not have mating seasons (as domesticated animals, and this behavior would likely continue into the apocalypse for generations), so breeding is active all year round. Secondly, litter size is averaged since randomized litter sizes isn't a thing (yet), and approximated full growth at 9 months (canines typically reach adulthood at 1 year, but they have intermediate juvenile stage that varies, but generally starts at 6 months and ends at 1 year, but they should be capable of defending for themselves especially as pack animals) --- which their puppy stats are abysmal, so wild puppies will probably be extremely rare, but quite capable if protected by player or placed in kennels.

Flag Rationale:
~~BITE - Over 15" height~~
CLIMBS - Same as LEAP requirements for now -- dog breed should be capable of high jumping without requiring specific training (essentially breeds that are known for jumping fences naturally)
~~FBITE - Unique to Pit Bulls - Ferocious Bite~~
GROUP_MORALE - Affection >= 4 and Friendliness >= 4
GUILT - Under 15" height
HARDTOSHOOT - Under 15" height, energy >= 4, > 100 speed
HIT_AND_RUN - Lost if Training >= 6 and Protection >= 6
LEAP - Energy >= 6 and Exercise >= 6
LUNGE - Training >= 4, Protection >= 4
~~MINORBITE - Under 15" height Shorter grabbed duration, less damage~~
SIZE:
- MEDIUM - Breed average over 21" and 83 lbs
- SMALL - Default size
- TINY - Breed average under 10" and 10 lbs

STUMBLES - Under 10" height, over 20lb weight
SWARMS - Friendliness >= 4

Answer for previous PR:
I believe this [source](https://www.petfinder.com/dog-breeds) will suffice for any future breeds as it has an extensive list that feels accurate.

Future tasks:

- Add back unique special attacks for dog breeds that performs bleed and grab on limbs (ignoring torso and head for now) as [specified here](https://github.com/CleverRaven/Cataclysm-DDA/pull/23813/files#diff-03ab30d902879795c572e282f23d3885).

edit: Other long-term goals (anyone welcome to work on):

- Expand on taming dogs (that allow the player more control over the pets, such as the dog whistle allows and increased interaction to determine the actions for the pets)

- Train dogs for certain roles (mainly as attack/guard/hunt animals, but additionally allow for bonding for morale / enjoyment purposes ["lap dogs"] -- give another reason for players to maintain stationary bases [also builds on acidia player bases, perhaps add kennels] rather than mobile vehicle bases)

### CODE REVIEW

**Comment by Inglonias** [comment]
I don't feel that this is... strictly necessary? Among other things, it makes it difficult to determine at a glance whether what I'm facing is dangerous, dog-wise. It feels like this is adding something just for the sake of adding it. What does this add to our gameplay experience?

**Comment by John-Candlebury** [comment]
Yeah maybe keep the ones that are strong opponents or are somehow unique, and remove the rest?

**Comment by 0be1isk** [comment]
> Among other things, it makes it difficult to determine at a glance whether what I'm facing is dangerous, dog-wise.

Unless you are a fresh spawn (and even that's being generous), there are almost no scenerios where a dog is dangerous.

> It feels like this is adding something just for the sake of adding it. What does this add to our gameplay experience?

I'd argue that it simply adds to the carrion system, which was well received. It wasn't strictly necessary to add such an implementation as all is does is over populate areas with bird feces, locusts, roaches, baby chicks, and other baby animals. I have yet to use any baby animal, but they are still there for the sake of being there as it just merely adds flavor to the game.

**Comment by Nymall** [comment]
As long as this doesn't lead to dog overcrowding in wildlife spawns, I don't see the harm. Some of the statistics I see surprise me(I live in southern BC and spent quite some time in Nova Scotia and Huskies are EVERYWHERE there), but all in all this is good for both variety and Role Playing.

**Comment by Jownky** [comment]
I don't know if what I'm going to say is a little bit off-topic, but as well as breeding them (Which I think is a great idea! :D), train (I've got the feeling that this has already been said) them to help you maybe bring you stuff or maybe send him to scavenge for any  useful items or even any sight of lifeform around the position of your PC.

**Comment by DracoGriffin** [comment]
>I don't feel that this is... strictly necessary? Among other things, it makes it difficult to determine at a glance whether what I'm facing is dangerous, dog-wise. It feels like this is adding something just for the sake of adding it. What does this add to our gameplay experience?

The descriptions for each breed could potentially use something different, but generally the dogs are similar enough in threat level (except maybe Chihuahua now that I think about it). I am open to suggestions if you would to like to review and offer them.

Not every PR is intended for gameplay, so much as it expands on the generic "dog" into something a bit more meaningful. The breeds that were chosen are extremely detailed if you refer to the original PR.

Also, some people like taming animals, some people like dogs, some people like to get attached to items or pets or vehicles or whatever other aspect. This allows that feature so people can have more detailed stories and interest.

>Yeah maybe keep the ones that are strong opponents or are somehow unique, and remove/mod the rest?

They are all pretty unique, as much as can be defined within the constraints between the game code and reality. If you have any suggestions on the formulas I developed, please let me know.

>As long as this doesn't lead to dog overcrowding in wildlife spawns, I don't see the harm. Some of the statistics I see surprise me(I live in southern BC and spent quite some time in Nova Scotia and Huskies are EVERYWHERE there), but all in all this is good for both variety and Role Playing.

That was my concern originally in the previous PR, and after extensive testing (playing with these changes locally since May), it shouldn't be any different than normal, except now regular dogs can be a different breed with differing pack sizes. Secondly, the statistics were just from shelters as those would be the most numerous in comparison to owner pets, especially since I only pulled data from the most populated city in each New England state (again, all of this is in extensive detail in the previous PR). Also, I didn't want to pay several grand for access to the pet owner statistics.

>I don't know if what I'm going to say is a little bit off-topic, but as well as breeding them (Which I think is a great idea! :D), train (I've got the feeling that this has already been said but as a fighting companion) them to help you maybe bring you stuff or maybe send him to scavenge for any useful items or even any sight of lifeform around the position of your PC.

Breeding is all thanks to Acidia and his carrion system. I had made a really bad json version but immediately switched to his. And yes, I'd love to have training, I have ideas, but my coding limitations is stuck with json, so if anyone is interested in developing training attack dogs, by all means, you have my full support and I'll be happy to provide whatever I can. Additionally, expanding on the "tame/pet" system would be amazing: I agree in perhaps you could train your dog for hunting to take down fowl or retrieve corpses and such. It'd also be nice to be able to play fetch or otherwise for some downtime (maybe even for NPCs at player faction bases). Moreso, expand it for robotic companions to make Eyebot take pictures of something to summon attack bots or tell a policebot to arrest a NPC, and so on. You should definitely make a separate Issue with some of these ideas of yours and hope someone can code them (if you can't yourself).

Lastly to the deleted comment, there actually is dog shit in this PR. It builds on the feces implemented in the carrion system in order to support and develop a food web (AKA food chain, food cycle) so wildlife would be a bit more procedural and capable of supporting itself instead of everything simply disappearing into nothingness. A great way to simulate life cycles to keep things interesting rather than static.

Otherwise, linted the mammal.json due to the errors but it makes it really ugly and more difficult to read/follow in my opinion. One more update to review smaller breed difficulty levels and this will be done unless any other issues arise.

**Comment by DracoGriffin** [comment]
Altered difficulty of select breeds to be half difficulty of mon_dog by any qualifying breed that has HP <50% of mon_dog's total HP (30, so anything below 15 HP has its difficulty reduced by half).

**Comment by Jownky** [comment]
This is anorther mere thought but some of them could spawn maybe in a Dog pound (I suppose it could be a new building?)? And there could be the posibility of there being none as well because maybe they escaped or even find the breeds zombified, these are just some posibilities. Keep up the great work! Excited to see detailed breeds, I personally want a German shepherd. :D

**Comment by DracoGriffin** [comment]
>This is anorther mere thought but some of them could spawn maybe in a Dog pound (I suppose it could be a new building?)? And there could be the posibility of there being none as well because maybe they escaped or even find the breeds zombified, these are just some posibilities.\

Yep, I actually have an animal pound (cramped, worse version than shelters, more chance of zombified, no rare breeds) and animal shelter (nicer version, open, rarer breeds -- may add rottweiler/great pyrenees to it). They can be previewed [here](https://github.com/DracoGriffin/DG_Mods/blob/master/DracoDogMod/mapgen/animalpound.json) and [here](https://github.com/DracoGriffin/DG_Mods/blob/master/DracoDogMod/mapgen/animalshelter.json). Will potentially be in a separate PR. I'm not sure how I feel about them as I particularly dislike mapmaking.

**Comment by DracoGriffin** [comment]
The two Travis build failures: 
`Terminated: data/mods//dda/../../json/monsters/mammal.json: line 892:12: value outside supported range` 

I'm not reproducing on my side when compiling, and debug spawning the associated monster(s) doesn't seem to produce any issues, and nothing is appearing in the debug.log about the issue.

Should I either ignore the failed builds or change the monster HP in some way to potentially rebuild Travis successfully?

**Comment by ZhilkinSerg** [comment]
> Should I either ignore the failed builds or change the monster HP in some way to potentially rebuild Travis successfully?

I will set hp of chihuahua to 6 before merging.

File: data/json/monsters/mammal.json [hunk 1]
```diff
@@ -855,7 +855,7 @@
     "description": "It's a tiny Chihuahua.  How it has managed to survive is a miracle; although it's small size and aggressive nature may been useful.",
```
Comment (Rivet-the-Zombie) [inline data/json/monsters/mammal.json]: >it's small size
Should read:
>its small size

Comment (DracoGriffin) [reply data/json/monsters/mammal.json]: Fix this in my next PR, thanks Rivet.

**Pull request merged**